shared/actions/login/saga.js

```diff
@@ -0,0 +1,588 @@
+// @flow
+// Look at this doc: https://goo.gl/7B6p4H
+
+import _ from 'lodash'
+import * as CommonConstants from '../../constants/common'
+import * as Constants from '../../constants/login'
+import * as DeviceConstants from '../../constants/devices'
+import * as Types from '../../constants/types/flow-types'
+import * as Creators from './creators'
+import HiddenString from '../../util/hidden-string'
+import engine from '../../engine'
+import {RPCError} from '../../util/errors'
+import {bootstrap, setInitialTab} from '../config'
+import {defaultModeForDeviceRoles} from './provision-helpers'
+import openURL from '../../util/open-url'
+import {devicesTab, loginTab, profileTab, isValidInitialTab} from '../../constants/tabs'
+import {isMobile} from '../../constants/platform'
+import {load as loadDevices} from '../devices'
+import {pathSelector, navigateTo, navigateAppend} from '../route-tree'
+import {overrideLoggedInTab} from '../../local-debug'
+import {toDeviceType} from '../../constants/types/more'
+import {call, put, take, race, select} from 'redux-saga/effects'
+import * as Saga from '../../util/saga'
+
+import type {DeviceRole} from '../../constants/login'
+import type {DeviceType} from '../../constants/types/more'
+import type {Dispatch, AsyncAction, TypedAction} from '../../constants/types/flux'
+import type {SagaGenerator, AfterSelect} from '../../constants/types/saga'
+import type {TypedState} from '../../constants/reducer'
+
+const deviceType: DeviceType = isMobile ? 'mobile' : 'desktop'
+const InputCancelError = {code: Types.ConstantsStatusCode.scinputcanceled, desc: 'Cancel Login'}
+
+const codePageSelector = ({login: {codePage}}: TypedState) => codePage
+
+function * generateQRCode () {
+  const codePage: AfterSelect<typeof codePageSelector> = yield select(codePageSelector)
+
+  if (codePage.textCode) {
+    yield put(Creators.setQRCode(codePage.textCode.stringValue()))
+  }
+}
+
+// TODO add waiting handlers
+// TODO sagaize
+const waitingForResponse = (waiting: boolean) : TypedAction<'login:waitingForResponse', boolean, void> => (
+  {payload: waiting, type: Constants.waitingForResponse}
+)
+
+const makeWaitingHandler = (dispatch: Dispatch): {waitingHandler: (waiting: boolean) => void} => (
```

Comment:
This was a bit hard to read at first -- hard to re-read it a few times to wrap my head around where the type definition was.